runtime/lua/vim/func/_memoize.lua

```diff
@@ -33,31 +35,67 @@ local function resolve_hash(hash)
   return hash
 end
 
+--- Memoizes a function {fn} using {hash} to hash the arguments.
+---
+--- Internally uses a |lua-weaktable| to cache the results of {fn} meaning the
+--- cache will be invalidated whenever Lua does garbage collection.
+---
+--- The memoized function returns shared references so be wary about
+--- mutating return values.
+---
 --- @generic F: function
---- @param hash integer|string|fun(...): any
---- @param fn F
---- @param strong? boolean
---- @return F
-return function(hash, fn, strong)
+--- @param hash integer|string|function Hash function to create a hash to use as a key to
+---     store results. Possible values:
+---     - When integer, refers to the index of an argument of {fn} to hash.
+---     This argument can have any type.
+---     - When function, is evaluated using the same arguments passed to {fn}.
+---     - When `concat`, the hash is determined by string concatenating all the
+---     arguments passed to {fn}.
+---     - When `concat-n`, the hash is determined by string concatenating the
+---     first n arguments passed to {fn}.
+---
+--- @param fn F Function to memoize.
+--- @param strong? boolean Do not use a weak table
+--- @return F # Memoized version of {fn}
+M._init = function(hash, fn, strong)
   vim.validate({
     hash = { hash, { 'number', 'string', 'function' } },
     fn = { fn, 'function' },
+    strong = { strong, 'boolean', true },
   })
 
-  ---@type table<any,table<any,any>>
+  hash = resolve_hash(hash)
+
   local cache = {}
+  ---@type metatable
+  local mt = {
+    __call = function(self, ...)
+      local key = hash(...)
+      if rawget(self, key) == nil then
+        rawset(self, key, vim.F.pack_len(fn(...)))
+      end
+
+      return vim.F.unpack_len(rawget(self, key))
+    end,
+    invalidate = function(self, ...)
+      local key = hash(...)
+      rawset(self, key, nil)
+    end,
+  }
+
   if not strong then
-    setmetatable(cache, { __mode = 'kv' })
+    mt.__mode = 'kv'
   end
 
-  hash = resolve_hash(hash)
+  setmetatable(cache, mt)
 
-  return function(...)
-    local key = hash(...)
-    if cache[key] == nil then
-      cache[key] = vim.F.pack_len(fn(...))
-    end
+  return cache
+end
 
-    return vim.F.unpack_len(cache[key])
-  end
+--- Invalidates the memoized function result for the given inputs.
+--- @param memoized_func function
+M._invalidate = function(memoized_func, ...)
+  getmetatable(memoized_func).invalidate(memoized_func, ...)
```

Comment:
Instead of using `getmetatable`, how about returning an object with an `invalidate` method? This would also mean not returning `cache` directly and doing something like:

```lua

  local obj = { 
    _cache = {},
    _hash = hash,
    invalidate = function(self, ...)
      local key = self._hash(...)
      self._cache[key] = nil
    end,
  }

  return setmetatable(obj, {
    __mode == strong and 'kv' or nil,
    __call = function(self, ...)
      local key = self._hash(...)
      local cache = self._cache
      if cache[key] == nil then
        cache[key] = vim.F.pack_len(fn(...))
      end
      return vim.F.unpack_len(cache[key])
    end,
  }
```

This would mean none of the call sites need to change.

Reply:
Nice, that's a good change